software/contrib/master_clock.py

```diff
@@ -0,0 +1,320 @@
+        self.MAX_BPM = 240
+        self.MIN_PULSE_WIDTH = 8
+        self.resetTimeout = 2000
+        self.previousStepTime = 0
+        self.minAnalogInputVoltage = 0.9
+        self.screen = 2
+        self.configMode = False
+        self.k2Unlocked = False
+        self.previousSelectedDivision = 0
+        self.previousActiveOption = ''
+
+        self.MAX_DIVISION = 128
+        self.clockDivisions = []
+        for n in range(self.MAX_DIVISION+1):
+            self.clockDivisions.append(n)
+        
+        # When enabled, set msDriftCompensation to 30
+        # When disabled, set msDriftCompensation to 28
+        self.DEBUG = False
+
+        # In testing a 17ms drift was found at all tempos and a 25ms drift when editing BPM/PW
+        # Adding this offset brings the BPM and pulse width back into a reasonable tolerance
+        self.msDriftCompensation = 17
+        self.msDriftCompensationConfigMode = 6
+
+        # Vars to drive UI
+        self.markerPositions = [ [0, 0], [69, 0], [0, 12], [40, 12], [80, 12], [0, 24], [40, 24], [80, 24]]
+        self.activeOption = 1
+
+        # Get working vars
+        self.loadState()
+        self.calcSleepTime()
+        self.getPulseWidth()
+
+        # Get asyncio event loop object
+        self.el = asyncio.get_event_loop()
+
+        # Starts/Stops the master clock
+        @b1.handler_falling
+        def StartStop():
+            self.running = not self.running
+
+        # Cycle screen and toggle config mode
+        @b2.handler_falling
+        def cycleScreen():
+            if ticks_diff(ticks_ms(), b2.last_pressed()) > 500 and ticks_diff(ticks_ms(), b2.last_pressed()) < 4000:
+                self.configMode = not self.configMode
+                # [temp] keep the clock running when in config mode to see if the previous slowdown bug occurs in testing
+                #self.running = False
+                if not self.configMode:
+                    # config mode has just been turned off, save state and lock k2
+                    self.saveState()
+                    self.k2Unlocked = False
+            else:
+                self.k2Unlocked = False
+                
+                # Turn off config mode to avoid current knob positions messing up other settings on the next screen
+                if self.configMode:
+                    self.configMode = False
+                    self.saveState()
+
+                if self.screen == 1:
+                    self.screen = 2
+                elif self.screen == 2:
+                    self.screen = 1
+                else:
+                    self.screen = 1
+
+        # Reset step count upon receiving a din voltage
+        @din.handler
+        def resetDin():
+            self.step = 1
+
+    '''Show running status'''
+    def screen1(self):
+        oled.fill(0)
+        oled.text(str(self.completedCycles) + ':' + str(self.step), 0, 0, 1)
+        if not self.running:
+            oled.text('B1:Start', 0, 23, 1)
```

Comment:
[optional] In my opinion, showing current state is more intuitive than the next toggle state.

Reply:
Did you test this using the very latest commit?
I did experience some of these bugs on previous versions...

Reply:
Updated known issues in the python file header and the markdown file.

Reply:
I figured out the config issue I was having. It looks like it's an issue with the min analog voltage on my EuroPi. The Master Clock was defaulting at 22 bpm, and any time I left Config mode, it reverted back to 22 bpm. Digging around more it looks like my ain voltage when unconnected was exceeding the `0.9v` min threshold, thus interpreting it as an input value. This is an excellent use case for #163 to configure each device's min ain threshold when its ready. In the mean time, I'd suggest making this a constant at the top of the script with some comments to make it easy for folks to adjust, or at least documenting this potential issue clearly.

Reply:
Thanks so much for taking the time to find the issue! I have capitalized and renamed the param to self.MIN_AIN_VOLTAGE and increased it to 1.1V. What voltage does your EuroPi fluctuate up to? Will 1.1V cover it?

Reply:
I guess its a good thing my development EuroPi has a misbehaving ain, because it has caught issues with other scripts too (https://github.com/Allen-Synthesis/EuroPi/pull/114#pullrequestreview-929470613)! When I measured the idle voltage then it was about `0.93v` which is _just_ over the threshold you had set.

I wanted to clarify a little bit more on what I meant by script constants. Typically, we don't want users to change any values inside the script classes or functions. However, sometimes there may be values that make sense for users to change to fit their personal needs. In those cases we want to provide constant definitions, ideally with some descriptions, at the very top of the script that users can change. That helps organize and group the user changeable values and does not require users to search through the code to find values they can tweak. Here are some examples implemented this way: ([example 1](https://github.com/Allen-Synthesis/EuroPi/blob/main/software/contrib/harmonic_lfos.py#L9), [example 2](https://github.com/Allen-Synthesis/EuroPi/blob/main/software/contrib/smooth_random_voltages.py#L56-L58), [example 3](https://github.com/Allen-Synthesis/EuroPi/blob/main/software/contrib/turing_machine.py#L41-L48). I'd suggest following this pattern and putting those constants at the top of the script to help users know they are values that can be tweaked to their preference.

Reply:
One more bit of discussion; I wanted to share my observations from measuring the clock using the Mordax DATA (https://imgur.com/a/rSACNfc). I tested running the Master Clock at 75, 125, 192, and 240 BPM to illustrate the decrease in stability and accuracy as the tempo increases. Folks familiar with Python should not be surprised by this (insert C fast / Python slow joke), but others might not expect this. Given that we're supporting a broad set of users I would encourage adding a note about this in the Known Issues section.

Side note, if you have any tips for me how I can record my EuroPi so the OLED isn't blurry, I'd love the help! I'm recording on a Pixel 6 pro, so maybe it has some "smart" photo enhancements going on that cause it to be blurry.

Reply:
Thanks for sharing this info. I will mod the script to include the constants using the method you laid out.

Regarding the BPM drift, I was not previously aware you were hoping for accurate BPM when using ain. This was never the intention of the feature because ain is so noisy, it is unlikely it could be used for accurate BPM control without a bit of code to try and average out the noise. However, as this feature was intended for wonky BPM and not accurate control, I will instead update the doc.

Any chance you could do the same BPM drift test without using the ain? I'm curious how accurate it is on your system.